protocol/contracts/templegold/TempleTeleporter.sol

```diff
@@ -0,0 +1,108 @@
+pragma solidity ^0.8.20;
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Temple (templegold/TempleTeleporter.sol)
+
+
+import { CommonEventsAndErrors } from "contracts/common/CommonEventsAndErrors.sol";
+import { ITempleERC20Token } from "contracts/interfaces/core/ITempleERC20Token.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { OApp } from "@layerzerolabs/lz-evm-oapp-v2/contracts/oapp/OApp.sol";
+import { MessagingReceipt, MessagingFee } from "@layerzerolabs/lz-evm-oapp-v2/contracts/oapp/OAppSender.sol";
+import { Origin } from "@layerzerolabs/lz-evm-oapp-v2/contracts/oapp/interfaces/IOAppReceiver.sol";
+import { ITempleTeleporter } from "contracts/interfaces/templegold/ITempleTeleporter.sol";
+import { OFTMsgCodec } from "@layerzerolabs/lz-evm-oapp-v2/contracts/oft/libs/OFTMsgCodec.sol";
+
+contract TempleTeleporter is ITempleTeleporter, OApp {
+    using OFTMsgCodec for address;
+
+    /// @notice Temple token
+    ITempleERC20Token public immutable override temple;
+
+    constructor(
+        address _executor,
+        address _temple,
+        address _endpoint
+    ) Ownable(_executor) OApp(_endpoint, _executor){
+        temple = ITempleERC20Token(_temple);
+    }
+
+    /**
+     * @notice Teleport temple tokens cross chain
+     * @dev Temple tokens are burned from source chain and minted on destination chain
+     * @param dstEid Destination chain id
+     * @param to Recipient
+     * @param amount Amount of tokens
+     * @param options LZ extra options
+     */
+    function teleport(
+        uint32 dstEid,
+        address to,
+        uint256 amount,
+        bytes calldata options
+    ) external payable override returns(MessagingReceipt memory receipt) {
+        if (amount == 0) { revert CommonEventsAndErrors.ExpectedNonZero(); }
+        if (to == address(0)) { revert CommonEventsAndErrors.InvalidAddress(); }
+        // Encodes the message before invoking _lzSend.
+        bytes memory _payload = abi.encodePacked(to.addressToBytes32(), amount);
+        // debit
+        temple.burnFrom(msg.sender, amount);
```

Comment:
In order to get the total circulating supply of Temple, would one have to sum up the totalSupply() on each chain?

Reply:
yes